### PR TITLE
[generate] Never use `cache_position` anymore in generation

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title. This is the last of many PR to remove the `cache_position`. At this point, all the models were already updated to not use them, and they are fully ignored in all the modelings. So this removes their creation and usage in `generate`, so they are not passed as `kwarg` anywhere anymore.
This is fully safe as all the models already ignore them.

Note: We keep a BC path for remote code models only, so that we still forward `cache_position` to the model. However, we don't maintain them in `model_kwargs` anymore, simply recreate them on-the-fly in `prepare_inputs_for_generation` if the model is remote code.

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
While I still remember it, let's kick it out from docs as well pls and if needed add correct examples with `cache`

**Comment by Cyrilvallez** [comment]
run-slow: dia

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -2068,9 +2068,6 @@ def set_model(self, model):
         self.model = model
 
     def set_inputs(self, inputs):
-        # build `cache_position` on the fly
-        seq_length = inputs["input_ids"].shape[1]
-        inputs = self.model._get_initial_cache_position(seq_length, self.model.device, inputs)
```
Comment (vasqu) [inline src/transformers/generation/logits_process.py:2073]: Just for my mind, can we run-slow with whisper

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -930,16 +923,6 @@ def _update_model_kwargs_for_generation(
...
 
-        # Cache position (always 1D)
-        if (cache_position := model_kwargs.get("cache_position")) is not None:
-            next_cache_position = (
-                torch.arange(num_new_tokens, dtype=cache_position.dtype, device=cache_position.device)
-                + cache_position[-1]
-                + 1
-            )
-            next_cache_position = torch.cat((cache_position, next_cache_position))
-            model_kwargs["cache_position"] = next_cache_position
```
Comment (vasqu) [inline src/transformers/generation/utils.py:941]: I think complete removal for cache position will be too breaking for remote code and there is still quite a lot, just looking at all the vLLM stuff we had to fix 😭 

Shouldn't pos ids be the same as cache positions now? What do you think about passing this as alias kwarg as well - we really need to check with a remote model, e.g. deepseek v3 remote code maybe?

File: src/transformers/integrations/executorch.py
Comment (vasqu) [inline src/transformers/integrations/executorch.py:1]: General question: Are we deprecating it everywhere?

I think I still see a few ocurrences:
- Mask creation
- Within this executorch integration
- Models
  - Lfm2
  - Ministral3
  - Mistral4
- Tests

Imo, only the mask might be critical and might be kept a bit longer. Wdyt?

**Review by vasqu — COMMENTED** [review]
Not approving yet because I want to discuss the deprecation a bit more
1. I still found references where imo they shouldnt be there
2. Do we keep cache positions in the generate preparation (alias for position ids): I think we will get a remote model apocalypse otherwise and vLLM already showed how brittle it is

File: src/transformers/integrations/executorch.py
Comment (Cyrilvallez) [reply src/transformers/integrations/executorch.py:1]: The remaining occurences are basically only a bit of BC! The breakdown is the following:

- mask creation: only there for BC
- executorch: it still needs `cache_position` because it needs to know when to reset the cache once the module was already exported. I will investigate later if we can use s simple `bool` for this instead, but not sure we can export with it then, and did not want to break the API in the meantime!
- models: it's just temporarry variables, not arguments -> could be renamed but no harm
- tests: only for executorch tests (see above!)

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -580,14 +577,18 @@ def prepare_inputs_for_generation(
...
 
-        # 6. Remove unexpected `generate` inputs (TODO @joao: fix trainer and examples)
-        model_inputs.pop("labels", None)
-        model_inputs.pop("next_sequence_length", None)  # if the method is overriden, usually it becomes a kwarg
+        # BC for remote code models only: create `cache_position` on the fly here, as we don't want to maintain them in kwargs
+        # between `forward`s
+        if self.is_remote_code():
+            past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
+            cache_position = torch.arange(sequence_length, device=input_ids.device) + past_seen_tokens
+            model_inputs["cache_position"] = cache_position
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py]: maybe a small deprecation warning, hopefully we can delete this workaround before v6 😆

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -862,11 +863,7 @@ def _expand_inputs_for_generation(
 
         def _expand_dict_for_generation(dict_to_expand):
             for key in dict_to_expand:
-                if (
-                    key != "cache_position"
-                    and dict_to_expand[key] is not None
-                    and isinstance(dict_to_expand[key], torch.Tensor)
-                ):
+                if dict_to_expand[key] is not None and isinstance(dict_to_expand[key], torch.Tensor):
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py:871]: hmm this is defi break custom generate repos (which I am not sure how many we have except for `dola/contrastive_search`)

File: docs/source/en/modular_transformers.md [hunk 5]
```diff
@@ -230,10 +229,8 @@ class Olmo2Attention(OlmoAttention):
         cos, sin = position_embeddings
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
 
-        if past_key_values is not None:
-            # sin and cos are specific to RoPE models; cache_position needed for the static cache
-            cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
-            key_states, value_states = past_key_values.update(key_states, value_states, self.layer_idx, cache_kwargs)
+        if past_key_values is not None:n}
```
Comment (vasqu) [inline docs/source/en/modular_transformers.md]:
```suggestion
        if past_key_values is not None}
```

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -580,14 +577,18 @@ def prepare_inputs_for_generation(
...
-            if key not in model_inputs:
+            if key not in model_inputs and key not in kwargs_to_avoid_forwarding:
                 model_inputs[key] = value
 
-        # 6. Remove unexpected `generate` inputs (TODO @joao: fix trainer and examples)
-        model_inputs.pop("labels", None)
-        model_inputs.pop("next_sequence_length", None)  # if the method is overriden, usually it becomes a kwarg
+        # BC for remote code models only: create `cache_position` on the fly here, as we don't want to maintain them in kwargs
+        # between `forward`s
+        if self.is_remote_code():
```
Comment (vasqu) [inline src/transformers/generation/utils.py]: Should we also check for the signature - might be a remote models adhering to the new changes in the future

File: src/transformers/integrations/executorch.py
Comment (vasqu) [reply src/transformers/integrations/executorch.py:1]: Would be nice just for the renames re the models - the others are good and we discussed enough internally 😂

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py:3734]: at some point in future, would be great to allow only new `inputs` for generation and deprecate the `cached+new` format. We're slicing it off anyway and afaik it doesn't bring any new information

**Review by vasqu — APPROVED** [review]
Really my last comments this time but looks good overall! Thanks for bearing with me 🫡

**Review by zucchini-nlp — APPROVED** [review]
Nice, happy to get rid of `cache_position` finally. I believe you already checked that all models have `kwargs` to consume `cache_position`, though a small test can be added to be sure

File: src/transformers/integrations/executorch.py
Comment (Cyrilvallez) [reply src/transformers/integrations/executorch.py:1]: Haha indeed! Done!

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/generation/utils.py]: Good point!

File: src/transformers/generation/utils.py
Comment (Cyrilvallez) [reply src/transformers/generation/utils.py:3734]: For sure, I'm planning some work there!

File: src/transformers/generation/utils.py (hunk 4 above)
Comment (Cyrilvallez) [reply src/transformers/generation/utils.py:871]: I don't think it's breaking, since any version below this one would create the cache_pos and this path, and any version after would not create cache pos and not check it either

**Comment by Cyrilvallez** [comment]
run-slow: dia, whisper
